### PR TITLE
Add image support for mod/modpack pages

### LINKED ISSUE #1015: download and show images in mod/modpack browser
### Role

i use polymc

### Suggestion

i'd like it to be able to download and display images in mod/modpack browser

### Benefit

better

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

image

this is what happens rn with images

### PR DESCRIPTION
**Pull request by flowln**
Closes #1015

I actually wanted to additionally add support for `<details>` and `<summary>` tags, since Modrinth's pages use those a lot, but I didn't find a way of doing so from outside Qt itself. Maybe someone smarter than me can figure it out :>

Questions that still need to be answered / things to be considered:
- Cache policy: Right now, it acts in the same way as other cached resources, always getting saved to the local cache after some time. However, those images can be quite big, so maybe we should have another cache policy for them. (We could also add a general feature in the cache system to limit a cache from a given meta entry to a certain size, and auto-evict old entries if the limit has been reached, kinda like how QPixmapCache works?)
- Options: Loading those images can be quite network-intensive, and specially on slower connections, make the application act sluggish-ly when entering a page with images not yet cached. We could perhaps have an option to not load those images, instead only removing the block that would be the image, so as to not have those blank page thingies. Where would those options be? I think the best place would be in a clog menu on some corner of the page widget, but idk yet how to implement that.

~One noteworthy detail: Sometimes, when images take too long to load, and since it started the user switched to a different page, the loaded image from the previous page will be inserted into the content of the new page, causing issues. It is resolved by simply switching pages again. Maybe it's obvious, but I didn't find the right signal to connect to get when the page changes, to perhaps abort ongoing requests :|~ fixed in [`d194b02`](https://github.com/PolyMC/PolyMC/pull/1126/commits/d194b02e28132df3ea3da961299e969614b8a185)

Also, some images may be misaligned, because of CSS and stuff (but at least we aren't using a webview! =D)

now, to the thing that anyone actually cares, the pretty photos:
plasmo voice with photos!
yes i chose it partially because its an anime girl, dont judge
_(issues with the icons on the list are unrelated to this PR and will be fixed in another one)_

### CODE REVIEW

**Comment by DioEgizio** [comment]
seems to work fine apart from what you already wrote, but plasmo voice has a few animated GIFs. Any chance of seeing them animated?

**Comment by flowln** [comment]
> seems to work fine apart from what you already wrote, but plasmo voice has a few animated GIFs. Any chance of seeing them animated?

not sure, after fetching the images we just throw them to Qt to do their thing, so if it didn't now, I don't think it will. Technically we could use a `QMovie` inside a `QLabel` to play a GIF, but inside the text document, I don't think we can have this sort of liberty, as it will only draw once per geometry / content update.

File: launcher/ui/widgets/VariableSizedImageObject.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+enum FormatProperties { ImageData = QTextFormat::UserProperty + 1 };
+
+QSizeF VariableSizedImageObject::intrinsicSize(QTextDocument* doc, int posInDocument, const QTextFormat& format)
+{
+    Q_UNUSED(posInDocument);
+
+    auto image = qvariant_cast<QImage>(format.property(ImageData));
+    auto size = image.size();
+
+    auto doc_width = doc->textWidth() - 2 * doc->documentMargin();
```
Comment (Scrumplex) [inline launcher/ui/widgets/VariableSizedImageObject.cpp:41]: why `- 2`?

**Review by Scrumplex — APPROVED** [review]
Looks good. Maybe add a comment where you are calculating `doc_width`. Also! Don't forget the license header!

File: launcher/ui/widgets/VariableSizedImageObject.cpp (hunk 1 above)
Comment (flowln) [reply launcher/ui/widgets/VariableSizedImageObject.cpp:41]: i added a comment for that, but just to close this, the reason is that `textWidth()` returns a width that includes the margins, so we need to remove those in order to have the image fit nicely with the rest of the text

**Comment by DioEgizio** [comment]
quickly switching between a page and another also causes to show the text of the old one. at least on mod downloading, haven't seen this on the modpack one

**Comment by DioEgizio** [comment]
also, might make sense to apply this to the news reader

**Comment by flowln** [comment]
> quickly switching between a page and another also causes to show the text of the old one. at least on mod downloading, haven't seen this on the modpack one

I've investigated the images issue i mentioned in the description, and was able to find a solution. However, it seems that the text issue you mentioned aren't directly related to what i've done in this PR. Trying on a develop build, i got a similar behavior, and in particular, it seems donate links sometimes get duplicated o.O

I'm not sure what's causing it, but it seems to be something to do with the async requests overriding one another in some cases (?)

**Pull request closed without merging**